ui/jetbrains/src/main/java/com/neueda/jetbrains/plugin/graphdb/jetbrains/ui/datasource/interactions/DataSourceDialog.java

```diff
@@ -35,51 +41,93 @@ public boolean go() {
+                .setTitle("Test connection")
+                .createPopup()
+                .showInCenterOf(contentPanel);
+    }
+
+    private void validateConnection(JPanel popupPanel, JComponent contentPanel, JButton testConnectionButton, JPanel loadingPanel, AsyncProcessIcon loadingIcon) {
+        ExecutorService executorService = ServiceManager.getService(ExecutorService.class);
+        showLoading(testConnectionButton, loadingPanel, loadingIcon);
+        executorService.runInBackground(
+                this::executeOkQuery,
+                (status) -> connectionSuccessful(popupPanel, contentPanel, testConnectionButton, loadingPanel, loadingIcon),
+                (exception) -> connectionFailed(exception, popupPanel, contentPanel, testConnectionButton, loadingPanel, loadingIcon),
+                ModalityState.current()
+        );
+    }
+
+    private String executeOkQuery() {
+        DataSourceApi dataSource = constructDataSource();
+        DatabaseManagerService databaseManager = ServiceManager.getService(DatabaseManagerService.class);
+        GraphDatabaseApi db = databaseManager.getDatabaseFor(dataSource);
+        GraphQueryResult result = db.execute("RETURN 'ok'");
+
+        Object value = result.getRows().get(0).getValue(result.getColumns().get(0));
+
+        if (value.equals("ok")) {
+            return "ok";
+        } else {
+            throw new RuntimeException("Unexpected test query output: " + value);
+        }
+    }
+
+    private void connectionSuccessful(JPanel popupPanel, JComponent contentPanel, JButton testConnectionButton, JPanel loadingPanel, AsyncProcessIcon loadingIcon) {
+        hideLoading(testConnectionButton, loadingPanel, loadingIcon);
+        JLabel connectionSuccessful = new JLabel("Connection successful!", AllIcons.Process.State.GreenOK, JLabel.LEFT);
+        popupPanel.add(connectionSuccessful, BorderLayout.CENTER);
+
+        createPopup(popupPanel, contentPanel);
+    }
+
+    private void connectionFailed(Exception exception, JPanel popupPanel, JComponent contentPanel, JButton testConnectionButton, JPanel loadingPanel, AsyncProcessIcon loadingIcon) {
+        hideLoading(testConnectionButton, loadingPanel, loadingIcon);
+
+        JLabel connectionFailed = new JLabel("Connection failed: " + exception.getMessage(), AllIcons.Process.State.RedExcl, JLabel.LEFT);
+
+        JTextArea exceptionCauses = new JTextArea();
+        exceptionCauses.setLineWrap(true);
+
+        Throwable cause = exception.getCause();
+        while (cause != null) {
+            exceptionCauses.append(cause.getMessage() + "\n");
+            cause = cause.getCause();
+        }
+
+        JBScrollPane scrollPane = new JBScrollPane(exceptionCauses);
+        scrollPane.setPreferredSize(new Dimension(-1, HEIGHT));
+        popupPanel.add(connectionFailed, BorderLayout.CENTER);
+        popupPanel.add(scrollPane, BorderLayout.SOUTH);
+
+        createPopup(popupPanel, contentPanel);
```

Comment:
Probably popup should not be shown if the connection dialog was closed before the check is completed.

Reply:
done